server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java

```diff
@@ -359,7 +407,7 @@ private KeywordFieldType buildFieldType(MapperBuilderContext context, FieldType
 
         @Override
         public KeywordFieldMapper build(MapperBuilderContext context) {
-            FieldType fieldtype = new FieldType(Defaults.FIELD_TYPE);
+            FieldType fieldtype = fieldType(indexSortConfig, indexMode, context.buildFullName(leafName()));
```

Comment:
Depending on `subobjects` the value of `leafName` could include the parent (before the dot) or not. We need the full name to check if we are dealing with `host.name` no matter the `subobjects` setting.